pkg/ddl/executor.go

```diff
@@ -10524,46 +10511,64 @@ func lockGlobalIDKey(ctx context.Context, ddlSe *sess.Session, txn kv.Transactio
 	return forUpdateTs, err
 }
 
-// combineBatchCreateTableJobs combine batch jobs to another batch jobs.
-// currently it only support combine CreateTable to CreateTables.
-func combineBatchCreateTableJobs(jobWs []*JobWrapper) ([]*JobWrapper, error) {
+// mergeCreateTableJobs merges CreateTable jobs to CreateTables.
+func mergeCreateTableJobs(jobWs []*JobWrapper) ([]*JobWrapper, error) {
 	if len(jobWs) <= 1 {
 		return jobWs, nil
 	}
-	var schemaName string
-	jobs := make([]*model.Job, 0, len(jobWs))
-	for i, jobW := range jobWs {
+	resJobWs := make([]*JobWrapper, 0, len(jobWs))
+	mergeableJobWs := make(map[string][]*JobWrapper, len(jobWs))
+	for _, jobW := range jobWs {
 		// we don't merge jobs with ID pre-allocated.
-		if jobW.Job.Type != model.ActionCreateTable || jobW.IDAllocated {
-			return jobWs, nil
+		if jobW.Type != model.ActionCreateTable || jobW.IDAllocated {
+			resJobWs = append(resJobWs, jobW)
+			continue
 		}
-		if i == 0 {
-			schemaName = jobW.Job.SchemaName
-		} else if jobW.Job.SchemaName != schemaName {
-			return jobWs, nil
+		// ActionCreateTables doesn't support foreign key now.
+		tbInfo, ok := jobW.Args[0].(*model.TableInfo)
+		if !ok || len(tbInfo.ForeignKeys) > 0 {
+			resJobWs = append(resJobWs, jobW)
+			continue
 		}
-		jobs = append(jobs, jobW.Job)
+		// CreateTables only support tables of same schema now.
```

Comment:
Why do we have this limitation?

Reply:
initial design has this limit https://github.com/pingcap/tidb/pull/28763

we want to extend it to support tables of different schema, but requires CDC to support it too, maybe in 8.4 or later